library/ecp_curves.c

```diff
@@ -836,7 +836,7 @@ int mbedtls_ecp_group_load( mbedtls_ecp_group *grp, mbedtls_ecp_group_id id )
 #endif /* MBEDTLS_ECP_DP_CURVE448_ENABLED */
 
         default:
-            mbedtls_ecp_group_free( grp );
+            grp->id = id;
```

Comment:
I don't understand why we need to set grp->id to id in the default case. Isn't that already its value?

Reply:
Hi @Patater ,

this is just zeroing, instead of full call of `mbedtls_ecp_group_free()` which does much more logic and calls slow and unnecessary `mbedtls_platform_zeroize()`. As @gilles-peskine-arm  note the `id` might be changed above in `grp->id = id;`, so the zeroing is required.

This particular call isn't a big problem, but we observed significant performance problems due to extra memory allocations, zeroing and copying https://github.com/tempesta-tech/tempesta/issues/1064#issuecomment-519182986 . We're still in progress of optimizations and will be able to backport the changes later.

This single-liner is just a small cleanup which can be easitly done.

Reply:
This shouldn't be `grp->id = id`, it should be `grp->id = 0`. The point is that in the error case, `grp` should to a blank state. (Remaining in a blank state is not practical since `grp->id` must be set to `id` in all other cases.)

But I'm not convinced that it's better to use `grp->id = 0` than `mbedtls_ecp_group_free`. Sure, calling this function takes a few more cycles, but it also makes the code harder to maintain.

Reply:
And regarding performance in general, Mbed TLS does call malloc more often than it should. But the main culprit is the bignum module. No allocations are involved here.

Reply:
Guys, I'm very sorry for the misprint - it's a big shame to make the single-line change only with 3 tries.

In our profiling we also observed memory zeroing among the most hot spots, but if you feel that the change is too minor improvement and doesn't make much sense, then feel free just to close the pull request.